### ISSUE TITLE
Remove element wise operations of exp and gamma to prevent confusion

### ISSUE THREAD

**Issue by gwhitney** [josdejong/mathjs#2440]
There is a standard meaning for the Gamma function applied to a square matrix (see https://arxiv.org/abs/1806.10554), which is not the Gamma function applied componentwise, like the current behavior of `gamma(M)` in mathjs.  Also, the componentwise application already has another very short expression, namely `map(M, gamma)`.

To avoid potential confusion, should the current matrix interpretation of `gamma(M)` as componentwise application be eliminated, requiring the use of `map(M, gamma)`? That way, no one would accidentally call `gamma` on a square matrix thinking they were getting the usual matrix gamma function. (And if anyone ever chose to implement the usual matrix gamma, it could either be overloaded onto the gamma symbol, or called `gammam` by analogy with `expm`.)

In fact, the analogy with `exp` is very strong. The usual mathematical meaning of "exp" of a (square) matrix is not componentwise exponentiation, but rather the matrix exponential. Clearly a specific decision was made to call that operation `expm` in mathjs, and I am not suggesting that decision be revisited. But I would suggest that the componentwise meaning of `exp(M)` be disabled and the documentation of `exp` be revised to point to `expm` for the matrix exponential and `map(M, exp)` for componentwise exponentiation. Again, the motivation here is to avoid the potential confusion of someone calling `exp(M)` thinking they were getting `expm(M)`.

Failing that, I'd suggest amplifying the documentation of exp to make it more prominent that `exp(M)` is not the matrix exponential.

In any case, perhaps it would be good to make a decision at least on whether `gamma` should continue to apply componentwise to matrices before acting on PR #2416 (which is the proximate cause of my looking into the semantics of `gamma` and raising this issue).

**Comment by josdejong** [comment]
Thanks for bringing this up. I wasn't aware of that, but it makes sense to remove the matrix support from both `gamma` and `exp` to prevent confusion. 

This would be a breaking change, so we will need to publish it as a new major version, which is fine (that is what the version numbering is for).

**Comment by gwhitney** [comment]
OK. Is there any particular protocol then with respect to a PR that makes this change, and/or any particular timing? Other interface-breaking items that are queued up that need to get in around the same time? Just let me know and I can make a PR for this; and also the PR in progress for `lgamma` (#2417) should be changed to not apply `lgamma` to every element of an array, so that we don't introduce another instance of this same sort of thing.

**Comment by josdejong** [comment]
There is not a specific protocol for breaking changes. If there are multiple breaking changes it's of course nice to bundle them in one go. So in this case we could bundle it with the breaking change in #2370 for example.

**Comment by josdejong** [comment]
> also the PR in progress for `lgamma` (#2417) should be changed to not apply `lgamma` to every element of an array, so that we don't introduce another instance of this same sort of thing

good point!

**Comment by gwhitney** [comment]
OK, I will put in a PR right away to remove the elementwise operation from gamma and exp so that you have flexibility for when to merge and generate a new major version. (If it sits for a bit because you want to wait on a major version bump, I totally understand.)

**Comment by gwhitney** [comment]
When I looked up the matrix exponential, it referred to the matrix logarithm and hyperbolic sine, which in turn referred to the matrix sine, etc. Basically any function defined by a simple power series has an analogue for square matrices that provides the usual definition of that function on square matrices. So the PR will eliminate all of them. If you want to draw the line somewhere else, just let me know and I will revise.

Incidentally, the current typing of say `asin` for TypeScript is that given a number it may return a Complex, unless config.predictable is true. Since predictable defaults to false, should I change the TypeScript typing of asin(x: number) to be `number | Complex` ?

**Comment by josdejong** [comment]
> Incidentally, the current typing of say asin for TypeScript is that given a number it may return a Complex, unless config.predictable is true. Since predictable defaults to false, should I change the TypeScript typing of asin(x: number) to be number | Complex ?

Yes that the correct TypeScript definition is `number | Complex` there (for example `sqrt` has something similar in that regard).

**Comment by gwhitney** [comment]
Yes, I guessed that's how you would respond, so in fact this change in typing is already in #2465.

**Comment by josdejong** [comment]
👍

**Comment by dvd101x** [comment]
Hi, first of all I’m constantly impressed with the capabilities of this library and constant development.

I would like to discuss the topic of an alternative to this decision.

There is a benefit for most of the functions to keep the capability of working with scalars and arrays as it allows for simpler evaluations. There are some examples of the workings of these functions in matlab and numpy using both scalars and N dimensional arrays.

- [exp numpy](https://numpy.org/doc/stable/reference/generated/numpy.exp.html)
- [exp matlab](https://www.mathworks.com/help/matlab/ref/exp.html)
- [Gamma numpy](https://scipython.com/book/chapter-8-scipy/examples/the-gamma-function/)
- [Gamma matlab](https://www.mathworks.com/help/matlab/ref/gamma.html)

As shown in the previous examples matlab and numpy can work on both functions as scalars and arrays.

It seems **mathjs** brings to *JavaScript* similar benefits to what **numpy** brings to *Python* surpassing it in some very impressive cases and in some cases it can borrow some ideas.

I have not used the function **gamma**, but the array like capabilities of **exp** seems as important as the array capabilities of **sin** or **cos**

**Comment by gwhitney** [comment]
Thanks for the references and the feedback. I don't see the matrix exponential in numpy, but I do see that matlab has `exp()` apply elementwise to matrices, and `expm()` for the matrix exponential. mathjs just makes a different choice, `map(M, exp)` to apply it elementwise, and `expm()` for the matrix exponential (if I am recalling correctly), so there is really no chance of being confused which one `exp(M)` does, since it doesn't do anything. And `map(M, exp)` is brief and very clear, so I am not seeing the problem you allude to? You can also do `map(M, gamma)` or `map(M, sin)` so it very easy to get array-like capabilities for any unary function, even one you write yourself.

**Comment by dvd101x** [comment]
Thanks for reviewing this and for the research on `expm( )` on other projects.

Even though numpy doesn't have `expm( )`, there is one in [scipy.linalg.expm](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.expm.html#scipy-linalg-expm). The same happens for `gamma( )` in [scipy.special.gamma](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.gamma.html#scipy-special-gamma).

I recently encountered one case where `exp( )` was expected to work for both scalars and arrays element wise. Thus a new function had to be done to work either way

``` python
# If the input is a scalar use exp(input), if it's an array use map(input, exp)
Exp(input) = equalText(typeOf(input), 'number') ? exp(input) : map(input, exp)
```

As you stated on other projects `exp( )` works for scalars and element wise to  matrices and `expm( )` for the matrix exponential. I understand that **mathjs** doesn't have to be like other projects and can make different choices.

I see that the same change has been happening for these functions 
* `exp( )`
* `expm1( )`
* `log( )`
* `gamma( )`
* `sqrt( )`
* `cbrt( )` 

I did some quick tests on these functions with numpy/scipy and Matlab and it's the same difference for all. It's expected to work for scalars and element wise for matrices and arrays.

I understand this is to avoid some confusion with matrix operations, nonetheless  I prefer the element wise option available on numpy/scipy and Matlab as it's very convenient to avoid some loops or maps.

I think I was writing my opinion too categorically, please note that I respect your decision and I will work around these functions that do not support matrices and arrays anymore.

**Comment by dvd101x** [comment]
Hi Jos, thank you for looking into it and reviewing the examples!

I'm very used to vectorization in [Matlab](https://www.mathworks.com/help/matlab/matlab_prog/vectorization.html) and [Octave](https://docs.octave.org/v4.2.2/Basic-Vectorization.html). I understand mathjs can be different and I'm wrapping my head around it.

For functions that can take two arguments like `log(x, base)` I guess the mapping would need an intermediate function for map to work as I don't think we can map with two arguments.

**Comment by gwhitney** [comment]
Yes, that is a good point; I don't think mathjs has a lambda-expression notation at the moment. Perhaps it ought to. Or, as less of a change, we could consider reverting just log(M, base) to operating elementwise, since although there is a standard operation of taking the log of a matrix, it's just to the natural base.  I.e., you wouldn't be specifying a base if you meant the matrix log, which would free up the semantics of log(M, base) to mean elementwise. But that exception might be a bit confusing, and certainly a lambda expression so you could say `map(M, x=>log(x,2))` or something like that would be more generally useful.

**Comment by dvd101x** [comment]
Hi Glen, I agree it would be nice to have lambda-expressions.

The shorter notation I found, that currently works on mathjs is `map(M, f(x) = log(x, 2))` and I think the scope of `f( )` is local.

As a reference a Matlab like notation would be `@(x) log(x, 2)` instead of `f(x) = log(x,2)` or `x=>log(x,2)`. Even though on Matlab there is no direct way of specifying the base for `log( )`

**Comment by gwhitney** [comment]
Sure, or since we now live in the unicode age we could have lambda-expressions literally be `λx log(x,2)` which is where the `@` in the matlab notation comes from anyway, as a stand in for lambda. But this is math**js** after all, so it seems staying close to the JavaScript notation for lambda expressions would make sense, hence `x => log(x,2)`, and then we could retcon `f(x) = log(x,2)` as syntactic sugar for `f = (x => log(x,2))`. But any notation is fine, and the fact that `map(M, f(x) = log(x,2))` currently works is very nice and means there's not much urgency in adding anonymous lambda expressions, since we already have named lambda expressions working.

**Comment by dvd101x** [comment]
I like what you propose `λx log(x,2)` including the possibility of multiple inputs. Nonetheless any notation is good and I trust there there will be a mathjs way of doing things best.

In the meantime I think for cases like this I will use `map(1:3, _(x) = log(x, 2))` which currently works and I associate `_( )` with anonymous.

I expected that on previous versions like **10.6.4** this could be written as `log(1:3, 2)` but that wasn’t the case either, maybe it was a bug.

**Comment by josdejong** [comment]
Interesting thoughts.

So we now have two discussions I think:

**1) lambda syntax**

See if we can improve on the syntax to write lambdas. The examples I see are:

```
map(M, f(x) = log(x, 2)) # this works right now

map(M, x => log(x, 2))
@(x) log(x, 2)
λx log(x, 2)
```
And I would like to add the following option to the list:

```
map(M, log(x, 2))
``` 
This last one can be confusing (because it compiles `log(x,2)` different depending on the context) and very clear (it reads very straightforward) at the same time. That can only be supported by the expression parser though, it can't work in the plain JS API. I implemented something like that in https://mathnotepad.com, where you can simply enter `plot(log(x, 2))` to plot a chart.

At this point I have a slight preference for a JS like interface like `map(M, x => log(x, 2))`, but this doesn't add that much over the existing notation `map(M, f(x) = log(x, 2))`, so I'm not sure whether it is worth adding it for that reason.

**2) support for element-wise operations**

Back to the original discussion: @gwhitney what are you're thoughts on this? It's mostly about safety vs convenience I think. But how big is the risk? Applications like Matlab and Numpy do implement element-wise calculation of all these functions (both exp and trigo). David's points (and my own experience) make me rethink our earlier decision on this 🤔.

**Comment by gwhitney** [comment]
Well I was drawn to mathjs in the first place for two reasons: mathematical fidelity (for example, mod works like in math, not like in javascript) and different data types are just handled smoothly. So for square matrices, there are standard, clear meanings for the (natural) log of a matrix, or the gamma function of a matrix, etc. So what else should gamma(M) mean, especially since there is such a concise clear syntax for elementwise operations?  So yes I would prefer not to use elemenwise interpretation of log(M) because of the standard matrix log operation. But of course I am just one vote and I understand the decision must be made collectively.

**Comment by dvd101x** [comment]
## 1) lambda syntax

I just saw a project with a similar syntax to `map(M, x => log(x, 2))` including a nomenclature for multiple inputs like `map(M, base, (x1, x2) => log(x1, x2))`. I think for it to work with multiple inputs `map( )` could *broadcast* the inputs.

> **Reference:** [MatlabJS](https://github.com/VivekTRamamoorthy/MatlabJS)

## 2) support for element-wise operations

In my case, I enjoy **mathjs** applications for engineering and for many things it is the best option compared to other references I know.

A simple example is `(20:25) celsius to fahrenheit` even though it’s an operation that makes sense for scalars and could also be written as `map(20:25, _(x) = x celsius to fahrenheit)` to do it elementwise; it’s nice to have that convenience and readability.

I don’t have strong opinions on `expm( )`, `logm( )` as I have not used them, I only have some references to other projects.

I’m biased to work on a vectorized way due to other software.

**Comment by gwhitney** [comment]
> A simple example is `(20:25) celsius to fahrenheit`

Let's be clear I am only talking about those functions that have distinct matrix versions. There is no matrix operation of "celsius to fahrenheit", therefore it's perfectly reasonable for the matrix versions of such operations to work elementwise; many mathjs operations currently do work that way. Just the ones that have conflicting matrix definitions do not; that seems like a perfectly reasonable way to avoid confusion to me.

**Comment by dvd101x** [comment]
Thank you for the clarification, you are right

I was trying to show an example of something that is currently in mathjs but in my opinion has a similar logic to:

> `sin((0:10) deg)` even though it’s an operation that makes sense for scalars and could also be written as `map(0:10, _(x) = sin(x deg))` or `map((0:10) deg, sin)` to do it elementwise; it’s nice to have that convenience and readability.

**Comment by gwhitney** [comment]
You bring up an interesting point. As far as I know, the the matrix-specific meanings of all of the operators for which the elementwise semantics have been removed are only for square 2-dimensional matrices. But currently in mathjs, there is only one matrix type, that handles n-dimensional matrices (n=1 ~ vectors, n=2 ~ typical rectangular matrices, n = 3 ~ 3D tensors -- somewhat less common, etc.)

I don't think there's any problem or ambiguity in basically allowing all unary functions to operate on vectors elementwise (although there might be cases I am not thinking of), but on the other hand, it would be weird to have operations behave differently on two different instances of the same data type. Is this perhaps an argument that vectors and matrices should actually be different data types? Or would it be just as problematic/confusing if `sin(V)` operated elementwise for vectors V but `sin(M)` either did nothing or computed the matrix sine for (square) matrices M?

**Comment by josdejong** [comment]
Thanks for your feedback Glen, I fully understand your point.

David, interesting to know that Julia has a mix of element-wise operation on scalars and matrix exponentiation for a square matrix. It feels confusing to me to have mixed behavior inside a single function.

Zooming out a bit: currently in mathjs we have a bit of an inconsistent mix:

- Functions like `add` and `subtract` evaluate element-wise
- Functions like `multiply` and `divide` evaluate in a mathematical way
- Functions like `exp`, `log`, `sin` do not support matrices and require using `map`. 

I love mathjs to be "mathematical sound", which is what Glen strives for. I also love mathjs to be practical, and that is currently lacking a bit: part of the functions do support element-wise operation, others don't. That's confusing. To me, having to use `map` introduces a serious barrier. I think having a more concise notation for element-wise operations would help me a lot. If there would be _no_ overhead/barrier, we could (very very maybe) even consider removing all element-wise operations to make the system consistent. Can we do a bit more brainstorming for a concise notation?

Thinking aloud here: for example for element-wise multiplication and division we have operators `.*` and `/*` (also used in Matlab for example). How about using this notation too on functions to execute element-wise? Like:

```
# element wise multiplication using a dot
[2,3] .* [4,5]

# ideas for a notation for element-wise operation of functions
[2,3] .* .log([2,3], 2)
[2,3] .* log.([2,3], 2)
[2,3] .* @log([2,3], 2)
```

**Comment by gwhitney** [comment]
I think you mean `.*` and `./` for the existing operators.

I think rather than considering this a systematic naming convention for element-wise functions on matrices (I know that such naming-convention approaches have not been a mathjs favorite in the past), what you are getting at is an operator that takes a a matrix and a function and applies the function elementwise to the matrix. In other words, deferring for a moment the question of what symbol to use for the operator by calling it `EltWise` temporarily, we would have that to take the sine of each element of a vector (1-d matrix) V, you would write either `V EltWise sin` or `sin EltWise V` or both would be legal. (Let's start just by thinking about unary functions.)

That seems pretty reasonable; you can in fact just think of `EltWise` as the infix version of `map`, which is nice, as it doesn't currently have an operator equivalent. As far as which argument order or both, the analogy with map suggests `V EltWise sin` but since the types lead to no ambiguity, for convenience I don't see any problem with allowing either order (and then for consistency I would suggest that map also allow either order if it doesn't already -- that would simplify implementation by precisely making EltWise the infix operator for map.

As far as what symbol should actually be used for EltWise, the analogy with `.*` and `./` does suggest `.` but unfortunately `.` is already used for field selection, so `V . sin` would become ambiguous between the `sin` field of object `V` and the function `sin` applied elementwise to `V`. So I don't see a way to use just a `.`.

APL has an "each" operator that does exactly this, and the Wikipedia entry for APL says the usual symbol for this operator is `¨` (a diaresis/umlaut). It's fairly suggestive, but not sure mathjs wants to go down the Unicode path?
An ASCII analogue might be `..` but as that's sometimes used for ranges it might be confusing used for `map`. On the other hand, mathjs already has a range operator and it's `:`, so at least nominally, the way is clear for `..` to be the map operator.  Supposing we allow either order of operands, that would give us either `sin .. V` or `V .. sin` as acceptable syntaxes for applying sin to each element of V producing a new vector. It looks pretty reasonable to me, so I'll advance that as a candidate although I'd certainly be open to other proposals.

OK, so that now brings us to binary functions, like taking the log base 2 of each element of a vector. From the above discussion, we already know  that with the `..` operator this could be done with `V .. _(x) = log(x,2)` (depending perhaps on precedence rules, maybe the right-hand operand would need parenthesization). With the adoption of other lambda-expression notations, this could be whittled down to something like `V .. x => log(x,2)`. To advance one more lambda syntax specifically for **unary** functions, since at the moment that's the only use case for the `..` operator, we could adopt the syntax `log(_, 2)` with the underscore being special here for the unary function that takes its argument to the log of that argument base 2 (i.e., `log(_,2)` is an abbreviation for lambda z log(z,2), where z is a fresh variable not occurring elsewhere in the expression. Then we would get `V .. log(_,2)` which is pretty concise.

There is one other possibility, if one is looking for this `..` to be as much like function application as possible. That is, if the function operand is multi-ary, then the other operand should be an n-dimensional matrix, and the size of its last dimension should equal the number of arguments that the function operand takes. Then the `..` operator would in this case return an (n-1) dimensional matrix where the values are the function applied to each slice along the last dimension. In other words, `log .. [[8, 2], [4,2], [16,2]]` would return `[log(8,2), log(4,2), log(16,2)]`, i.e., `[3, 2, 4]`. Then to get the effect of taking log base 2 of each element of a vector V, we would just need a quick notation for promoting the vector to a matrix by pairing each element with 2. There may already be such a notation (I have to wrap up this post, but I will look later), but let's imagine just for the moment that `(V, 2)` had that semantics (I am not saying that it does or could have that meaning without any conflict with current mathjs, but just imagine for the moment). Then with these conventions, `log .. (V, 2)` would mean precisely to take the log base 2 of each element of V, which would be about as short and sleek as I suppose you could get.

**Comment by josdejong** [comment]
That is a really interesting idea, introducing an operator such as `..` for the `map` operation. Or maybe we could name the operator `map` (just like we have `and` both as function and operator), than it can look like `V map sin` and `V map x => log(x,2)` (though it's not much more concise than what we have already). I have to give it a bit of thought 🤔

**Comment by gwhitney** [comment]
Right, at the moment only the logical operators can be used both as functions and as infix operators. If we open this up to map as well, we might be begging the question as to why you can't just use any binary function as an infix, e.g. `16 log 2` to get 4. Anyhow, if `map` is an infix operator, too, then additional brevity can be achieved with alternate notations that have been put forward, like `V map log(_,2)` or with the appropriate semantics `log map (V,2)`.

On that latter point, I checked that `V, 2` and `(V, 2)` are syntax errors and that the doc says that currently the only meaning of `,` is as a parameter separator. Therefore, it is at least plausible that some sort of matrix extension semantics could be given to `,` that would allow `log map (V, 2)` to work as outlined above (or `log .. (V,2)` if a the symbol `..` is used for the operator version of `map`). And the operation of filling in a constant to make a new column of a matrix is already pretty straightfowardly accessible in mathjs:
```
V = [27, 9, 81]
   [27, 9, 81]

resize(V, concat(size(V), [2]), 3)
   [[27, 3], [9, 3], [81, 3]]
```
So conceivably `V,3` could become a shorthand for this operation. And if you allow it for two vectors so that if  V is `[4, 9, 25]` and W is `[2, 3, 5]` then `V,W` becomes `[[4, 2], [9,3], [25,5]]` then with all of these conventions `log map (V,W)` would come to mean the log of each element of V to the base given by the corresponding element of W, or in this case, it would result in  `[2, 2, 2]`. Mostly just exploring the possible syntax/semantics here, but none of these options seem implausible to me...

**Comment by josdejong** [comment]
Ah, yeah, that is quite creative. So we could implement `(V,2)` as concat with broadcasting (see also https://github.com/josdejong/mathjs/discussions/2753#discussioncomment-4767925), and with `map` and `..` both supported as map operator, we could have a syntax like `log map (V, 2)` and `log .. (V,2)`. I think it is technically possible, but it feels tricky to me.

A syntax like `V map log(_, 2)` sounds more appealing to me. It is a bit similar to the special `end` variable like `A[2:end]`. It is more explicit what's going on I think. That is also very close to a function notation like `map(V, log(_, 2))`.

I just tried out  `map(V, log(_, 2))`, and that already works 😯😂 

Still, `map(V, log(_, 2))` or `V map log(_, 2))` is much more verbose than `log(V, 2)`. 
Only something like `log.(V, 2)` or `@log(V, 2)` would come close to it.

**Comment by josdejong** [comment]
@dvd101x what are your thoughts?

**Comment by gwhitney** [comment]
> Only something like `log.(V, 2)` or `@log(V, 2)` would come close to it

Slightly confused, why is `log .. (V, 2)` "tricky" but `log.(V, 2)` a good notation? They seem to differ only in a single dot... Or is it because you are thinking of `log.` as a function that is like log but applies to vectors, whereas the last proposal I made for `log .. (V,2)` involved expanding the right-hand operand into a two-D matrix, which seems a bit wasteful?

Alternatively if we allowed `(V, 2)` as simple pairing to produce a tuple whose first entry is V and second entry is 2, we could certainly create an operator, I will use the symbol `@` for now but it could be `..` or whatever fits the syntax/semantics of mathjs, so that `binaryFn @ (V,2)` applies the given binary function directly in a vectorized way to V with the second argument always 2. Then again `log @ (V,2)` would mean `[log(V[1], 2), log(V[2], 2), log(V[3], 2),....]` but without building the temporary matrix. Would that then not be "tricky" and provide a viable notation for this sort of vectorization?

**Comment by dvd101x** [comment]
I'm used to have different notations to divide *scalar or elementwise* vs *matrix* operations when they exist similar to `exp(A)` for *scalar or elementwise* and `expm(A)` for exponent of *scalar or NxN matirx*.

I'm glad you are looking into possibilities as I think using map adds some friction and could be avoided in some cases. Maybe a good tradeoff to keep some convenience and have better consistency is applying a similar syntax to the logic of `*`, add a dot somewhere for *scalar or elementwise* and otherwise it would work for *scalar or matrix*. It seems logical to have something that can play nice along `.*` and `./`

```python
x1 = 0:0.1:1
x2 = 1:-0.1:0
y = x1 .* x2 .* map(x1, exp) # current mathjs

# matlab reference x1 .* x2 .* exp(x1)
# a few of the previous proposals
# y = x1 .* x2 .* exp.(x1)
# y = x1 .* x2 .* .exp(x1)
```

Regarding the mapping of multiple inputs, it's something that I have been struggling with and might benefit from broadcasting 

```python
x = [2, 3]          # row vector
base = [[2], [3]]   # column vector
.log(x, base)       # would internally broadcast both inputs and then map both

# broadcasted x would become [[2, 3], [2, 3]]
# broadcasted base would become [[2, 2], [3, 3]]

# would yield
# [[1.0000, 1.5850],
#  [0.6309, 1.0000]]
```

And also map could use multiple inputs like.

```python
x = [2, 3]          # row vector
base = [[2], [3]]   # column vector
map(x, base, log)   # would internally broadcast both inputs and then map both

# broadcasted x would become [[2, 3], [2, 3]]
# broadcasted base would become [[2, 2], [3, 3]]

# would yield
# [[1.0000, 1.5850],
#  [0.6309, 1.0000]]
```
I'm showing examples of two ortogonal vectors converting to 2 dimensions, but the same logic applies to N dimensions

**Comment by josdejong** [comment]
@glen sorry, yes indeed I was thinking of `log.(V, 2)` or `@log(V, 2)` as a regular function call. It's indeed basically the same as `log..(V, 2)` if you look at the syntax. I like your idea of treating `(V, 2)` as a tuple, and have an operator like `@` or `..` to glue a lambda and a tuple with arguments together. Writing something like `log..(V,2)` or `log@(V,2)` looks quite ok to me. We have to be a bit careful how we document this and what we support exactly: the matrix `[V, 2]` is also a tuple. We should not get a mixed notation of angle and round brackets to denote matrices. I think there should be a clear limitation and separation in that you can only use `(V, 2)` for function arguments and not as a matrix.

@dvd101x thanks for your inputs, thinking too along the lines of something similar to the element wise operators `.*` and `./`. 

```
x1 .* x2 .* map(x1, exp) # current syntax
x1 .* x2 .* .exp.(x1)
x1 .* x2 .* exp(x1)
x1 .* x2 .* exp@(x1)
x1 .* x2 .* exp..(x1)
x1 .* x2 .* exp:(x1)
```
Thoughts:
1. Using a single dot `.` as prefix or suffix around the function name feels to me like a typo, and overloads the use of the `.` even further. 
2. Using `@` or `..` feels more like an intentional operation. Maybe `..` would be a good choice since it is nicely in line with `.*` and `./` operators and looks similar to the spread operator `...` which also communicates a "for each" message. Some languages like C# have `..` as range operator, and we could even consider supporting `1..5` to denote a range (current syntax: `1:5`). Having two syntaxes will be confusing though, I'm not sure if that would be a good idea. To make the syntax resonate with users, it would be helpful though if it comes back in multiple similar cases. 
3. Or, instead of `..`, we could consider `exp:(x1)`, which is then a similar syntax as a range `1:5` and still has two dots.

**Comment by gwhitney** [comment]
I have no issue with `:` as the operator symbol for this sort of vectorized function application. It doesn't seem to conflict with the existing semantics of `:` and indeed feels somewhat analogous to ranges, thinking of the latter as a sort of "vector" operation. And it even already associates right to left, so that `exp:1:4` would, as I think you'd expect, give the first four powers of e.

One point though I think that we should think about is whether there is any potential ambiguity/conflict with `:` as the key-value separator in object literals in the mathjs expression language. In other words, suppose `V` and `W` currently have values as vectors. Would `{exp:V, sin:W}` be confusing or ambiguous in its parse? Or are you thinking that that parentheses are a required part of the application notation, so that `exp:V` would not be allowed as a way of applying exp to every entry of V, only `exp:(V)`? So then the example mixed with range would have to be `exp:(1:4)` and if you wanted to apply say acos and then sin to a vector V of numbers you would have to write `sin:(acos:(V))` rather than `sin:acos:V`? I'm not advocating for either way on that, just exploring the details of the proposed notation.

**Comment by dvd101x** [comment]
Maybe the convention (to be defined) could also be useful for the user defined functions. I'm including a character that already works just as a reference to imply elementwise.

```python
f(x, y) = x ^ y
f([1, 2; 3, 4], 3)     # [[37, 54], [81, 118]]

$f(x, y) = x .^ y
$f([1, 2; 3, 4], 3)    # [[1, 8], [27, 64]]
```

Previously it was more or less the other way around for a few functions (like exp and expm)
```python
f(x, y) = x .^ y
f([1, 2; 3, 4], 3)      # [[1, 8], [27, 64]]

fm(x, y) = x ^ y
fm([1, 2; 3, 4], 3)     # [[37, 54], [81, 118]]
```

**Comment by josdejong** [comment]
A few days later, the idea of using `:` as a new elementwise operator still feels quite ok to me! It sort of makes sense to me (indeed being analogous to ranges).

@gwhitney , good point about possible conflicts with the key/value separator inside objects. This is technically possible though. It is similar to the `;` that already has two meanings: as separator between multiple expressions like `a=2;b=3;a+b`, and as row separator inside a matrix `[1,2;3,4]`. The parser can handle the operator differently depending on the context. The parser expects to parse a `:` right after parsing a key, and can keep doing that. And we can just add another operator in the chain to interpret `:` on other places.

I hadn't thought yet about making the parens optional, like `exp:1:4`, `exp:V`, and `sin:acos:V`. It makes sense to me though, it is really a clean and compact way of writing, that can truely "compete" with the old, no longer supported notations `exp(1:4)`, `exp(V)`, and `sin(acos(V))`. It is _more_ concise, _and_ conveys more information on what's going on. With `square(A)` and `exp(A)` you may be wondering how it is evaluated, but with `square:A` and `exp:A` you can directly tell.

> Maybe the convention (to be defined) could also be useful for the user defined functions.

Yes, definitely, and also for functions that you import from some other library and that are implemented for scalar values only. You can "automagically" use them elementwise on matrices using this new operator. 

Here a few more examples to get a feel for how things can work out:

```
# custom function
square(x) = x * x

A = [1,2,3,4]

# currently supported for elementwise evaluation:
map([1,2,3,4], square)
map(A, square)
map(1:4, square)
map([1,2,3,4], f(x) = log(x, 2))

# new syntax for elementwise evaluation:
square(A)     # 30
square:A      # [1, 4, 9, 16]
square:(A)    # [1, 4, 9, 16]
square:1:4    # [1, 4, 9, 16]
square:(1:4)  # [1, 4, 9, 16]
log:(1:4, 2)  # [0, 1, 1.58..., 2]
```

I quite like it. Shall we go for it?

**Comment by gwhitney** [comment]
> I quite like it. Shall we go for it?

Well, I think there are several questions/implications remaining here. Your comments have convinced me there is not a parsing/semantic ambiguity; the only other use of `:` is for the object key-value separator, and that only occurs inside `{ }`, and the only use of those braces is for objects, and so inside of braces, any top-level identifier is a key, and the colon after it is the separator from the value, and all other colons would then be this operator. So that's OK.

But here are some questions that remain:

1) Most importantly, to actually use this notation for anything other than unary functions, we need some sort of tuple entity to hold the multiple arguments. Are we contemplating a new tuple type, or re-use of an existing collection type like an array or a matrix (is there anything else besides those two)? 

1A) If the former (a new tuple type), then there is a lot to work out: the exact syntax/semantics for the new type, how it interacts with existing operators and with existing arrays/matrices, etc. And then if the syntax for a tuple of three things, say a number, a boolean, and a string, is `(2.5, true, "foo")`, then are we somehow reimagining function application `func(2.5,true,"foo")` as generation of a tuple object followed by function application of `func` to that tuple? In other words, juxtaposition is the multiplication operator when the left operand is a number and it is the function application operator when it is a function? Are we then allowed to write `abs 7` (right now you have to write `abs(7)` ) for a unary function like `abs`? Conversely, if `func` indeed takes three arguments as shown, are we now allowed to write `T = (2.5, true, "foo"); func T`? And in this latter case, would `func T` differ from `func(T)` (with `func T` meaning apply func to the tuple T as its three arguments, and `func(T)` meaning make a tuple with one element, where the one element is a tuple with three elements, and try to apply `func` to that in case it has a unary meaning where its single argument is allowed to be a tuple? I am not saying that's sensible, just exploring the logical consequences here... Or perhaps (i) there are no such things as tuples of one element, only tuples of two or more, and so anything that appears to be a tuple of one element is just that element itself, and/or (ii) maybe parentheses are idempotent on _all_ tuples, i.e. if `T` is any tuple at all, then `(T)` is just `T` back? (Note that both/either of (i) and (ii) are consistent with the general principle that for **any** `x` at all, `(x)` should just be `x` back, which is certainly the case now and seems like a good principle to adhere to...)

1B) If the other case (reusing arrays for tuples) then we need to take care that there aren't new ambiguities, and then probably we would write `log:[1:4, 2]` for your example above rather than `log:(1:4, 2)` because we wouldn't want two different notations for the array whose first element is the array `[1,2,3,4]` and whose second element is `2`. And then I think it might be tricky to tell, if you have some function `fun` that can be unary but can take a second optional argument, when `V` is a vector of two elements, whether `fun:V` is applying the binary interpretation of `fun` with the two elements of `V` as its arguments or applying the unary interpretation of `fun` to `V` elementwise. So it seems there would be details to work out in this case, too.

1C) Or should we go slowly and start with this just working for unary functions, where there's not really a need for tupling?

2) In any of these cases, there is the question of whether `:` can be used to just apply the function, i.e. is `exp:0` just 1 (or maybe `[1]`) or an error (i.e., maybe `:` can only be used in a vectorized context)? If `:` can be used in a non-vectorized context, how (if at all) does it differ from ordinary function application?

3) Every (not sure -- maybe just almost every?) existing operator is just the infix symbol for a named (prefix-notation) function. I don't think there is an existing named function that corresponds to this `:` operator. Should that be added first so its semantics can really be nailed down? What should it be called? It seems quite similar to `map`, but (a) a bit more general I think, and (b) with a different parameter order. Would `map` then be dropped in favor of this new function? Or could perhaps the existing `map` be (first?) generalized to cover all of these new cases so that `:` would in fact really be just the infix operator for `map`?

4) Your last example of `square` is interesting because there is already a built-in function square, and it doesn't work the way you show: `square(1:4)` is an error, precisely because it is ambiguous between returning `30` and returning `[1,4,9,16]`, so you must either write `(1:4)*(1.4)` or `map(1:4, square)` to disambiguate. Are you suggesting that after or as part of this change, we would make `square(x)` truly synonymous with `x*x` and rely on the new notation for elementwise operations when you want the `[1,4,9,16]` interpretation? Would we do away with auto-vectorizing functions altogether and just rely on the new notation? For example, even having removed the vectorized meaning of `gamma(x)`, it is still currently the case that `abs(-3:3)` returns `[3,2,1,0,1,2,3]` because it previously auto-vectorized, and there isn't particularly a meaning of the "absolute value" of a matrix, so it has just been left vectorizing. After this change, would `abs(-3:3)` be an error, so that if you want to do this, you have to write `abs:-3:3` (or possibly `abs:(-3:3)` if we make the parentheses mandatory)?

My guess is there will be other issues but those are the ones I can see at this sitting that should likely be considered/worked out before diving into an implementation. Seems like there is still a fair amount of ground to cover.

**Comment by josdejong** [comment]
Sorry for the confusion on `square` in my examples. I've renamed it to `mySquare` so it's clear we're not talking about the built-in `square` but just "some" unary function or custom function.

I think there is no need for introducing a new tuple kind of data structure to hold function arguments. I would like to avoid that because, like you explain, it brings a lot of complication. 

We can handle arguments on the right hand side of the `:` operator directly inside the parser as function arguments, as part of the `:` operator itself. Similar to regular function calls, which have a pattern of a symbol followed by arguments. There is no need for tuple data structure there either. Basically, we should extend the [parseRange](https://github.com/josdejong/mathjs/blob/develop/src/expression/parse.js#L936-L980) code in the parser such that it handles more cases than the ranges we have now.  

Implementation could roughly look something like this:

- We implement a new function named `elementwise(fn, arg1, arg2, arg3, ...)`, which evaluates as 
  `map(arg1, callback(x) => fn(x, arg2, arg3, ...))`. So, `elementwise(log, [1,2,3,4], 2)` is equivalent to `map([1,2,3,4], f(x) = log(x, 2))`.
- We implement a new node `ElementWiseFunctionNode`, which can hold a left hand side being a symbol or Node, and a right hand side being a list with function arguments. It is the same as `FunctionNode`, except that it invokes the new `elementwise` function on the function and arguments instead of invoking the function itself. Maybe we can extend/reuse `FunctionNode` for this, we will see.
- In `parseRange`, after parsing the first `:`, we add a new condition which checks for `(` being the next token. If so, we parse the next part as function arguments (maybe we can use/reuse `parseAccessors` there), and then we output an `ElementWiseFunctionNode`. This handles cases like:

    ```
    mySquare:(A)
    mySquare:(1:4)
    log:(1:4, 2)
    ```
- In `parseRange` we now have a new cases when creating a `new RangeNode(...)`. So far it supports `start:end` and `start:step:end`. But now, we can also have `fn:start:end` (like `mySquare:1:4`), `fn:start:step:end` (like `mySquare:0:5:100`) and `fn:arg1` (like `mySquare:A`). We have to extend `RangeNode` and function `range` such that they can handle element wise function calls, so it can call `elementwise(fn, range(...))`. The function you're referring to in your point 3 I think is the existing `range` function. This will handles cases like:

    ```
    mySquare:A
    mySquare:1:4
    range(mySquare, 1, 4)
    ```

**Comment by gwhitney** [comment]
OK, wow, I was not thinking about this clearly, sorry. If a proposal along these lines is adopted, then we want the meaning of `foo:bar:` to depend on whether `foo` is a function or a number. But the _parse_ of this expression can't depend on that, and it currently basically parses to `range(foo, bar)`, so we would have to extend the semantics of the `range` function to cover the case of `range(abs, [-3,7,-2])` doing elementwise application to produce `[3,7,2]`. That's a pretty huge expansion of the behavior of that function. Are you sure you are OK with that?

If so, then the fact that currently ranges accept optionally a third element as in `start:step:end` introduces wrinkles. One possibility is that we could use that notation for the binary, ternary, etc. function application cases. In other words, to take the log base 2 of each element of the vector V you would write `log:V:2` rather than `log:(V,2)`. So then `range(func, arg1, arg2, ..., argk)` would apply the k-ary function "func" to the arguments in turn generated by broadcasting all of arg1, arg2, etc, up to argk to the same shape`. And `range(number, number)` and `range(number, number, number)` would behave as they currently do, and `3:4:5:6` would parse to `range(3,4,5,6)` and would be a runtime error. Note that in this scheme, if you wanted to use a range in one of the positions of an elementwise application, you'd have to parenthesize it; in other words for the log of 1 through 4 base 2, you'd have to write `log:(1:4):2`, and you couldn't write `abs:-3:3` because that would end up trying to call `abs(-3,3)` which is not valid, you'd have to write `abs:(-3:3)`.

If you don't like that for binary vectorization, then yes, `log:(1:4,2)` could be a special syntax, but as far as I can see it might as well just parse directly to `range(log, 1:4, 2)` and still have the semantics of that be as above  -- I don't see any reason to create a new node type, etc.

In any case, since all of these `:` expressions have to parse into `range` (and so whether we want to use `:` for elementwise operations depends on whether you're comfortable extending the `range` function to these cases), I don't see that there's any need for a new `elementwise` function -- just new typed-function cases for the existing `range` function. (If we use a different operator symbol, then likely we'll need a new function for its semantics.)

So in other words, the `elementwise` function you write about should just be the new cases for `range`. But I don't think the semantics as described are quite right -- you might be vectorizing along some other argument or more than one simultaneously. In other words, you should be able to elementwise-apply `log` to `4` and `[2,3,4]` to get the log of 4 to each base 2, 3, and 4, or elementwise-apply `log` to `[4,9,16]` and `[2,3,4]` to get `[2,2,2]`. The expression you wrote doesn't cover these latter cases.

**Comment by josdejong** [comment]
I think you're right and we do not need an `elementwise` function at all, we can do with extending `range`. And we do not need a new `ElementWiseFunctionNode` but can just use `RangeNode`. Much simpler, Nice!

### Syntax

We could indeed consider interpreting `log:V:2` as  element-wise log with arguments V and 2, instead of `log:(V, 2)`. That would be technically simpler and more straightforward in a way. But I do like the similarity of the regular `log(V, 2)` and it's element-wise version `log:(V, 2)`. 

Mixing the original numeric range with the new element-wise function operator can introduce quite some confusion and weird cases. Maybe we should simply not allow `fn:start:end` and `fn:start:step:end`, and do not mix these two different things. So, we could do:

```
# numeric range
start:end
start:step:end

# element-wise function operator
fn:(arg1, arg2, arg3, ...)
fn:arg1             # shorthand for fn:(arg1) 

# do NOT allow mix of these two:
fn:start:end        # NOT supported
fn:start:step:end   # NOT supported

# function range
range(start, end)
range(start, step, end)
range(fn, arg1, arg2, arg3, ...)
```

It feels a bit weird indeed to extend `range` with an element-wise function call. But well, all these concepts of range, map, repeat are close together, and it makes sense to use the very same `:` operator for them. I have to say it looks good to me, but I'm still not totally convinced, have to let it sink in a bit. The alternative would be to use an other character, like `log@(V,2)`, or `log..(V,2)`, so these two keep cleanly separated (with a new `elementwise` function under the hood, separate from `range`). Or we could disallow the shortcut `fn:arg1` and require `fn:(...)`, so the parser can already recognize whether it's a numeric range or element-wise function.

### Philosophy

About your point (4) on `square`: it makes sense to me to implement `square` as `x*x` for matrices, and the same with `cube`. 
It still bothers me that there is an inconsistency in some functions like `exp`, `log`, `sin`, `sqrt` _not_ supporting matrices out of the box, and other functions like `abs`, `add`, `eqaul` `max` `sign` that _do_ support matrices. That can be confusing. Let me try to write out the philosophy we try to implement right now for functions operating on matrices:
- in principal, functions evaluate the mathematical way, like `multiply` and `divide`.
- if the mathematical way is element-wise evaluation, we do suppport that, like `add` and `subtract`.
- if the mathematical way is _not_ element-wise evaluation but we miss the mathematical implementation, we throw a clear error when invoking the function with a matrix, like `log`, `sin`, `square`. 
- If you want element-wise evaluation on any function, you can use the new explicit element-wise syntax like `log:(V, 2)`.

I'm not sure if there are still functions in a "gray" area in this regard, but I think `abs` is one of those. If so, we have to have a look at those. What do you think @gwhitney ?

**Comment by gwhitney** [comment]
Yes, I think we are homing in on the details here.  But I think there are still some parsing issues:

1) `foo:bar:baz:quux`

This is currently a syntax error, and if we are  not going to use this for applying a ternary function `foo`, it can remain a syntax error.

2) `foo:bar:baz` and `foo:(bar, baz)`

The first of these has to parse to a RangeNode with three children, in case foo, bar, and baz all mean numbers. So if the second form also parses to a RangeNode with three children, and the semantics of such a node is given by the `range(foo, bar, baz)` function, and if that means element-wise application of a binary function foo to bar and baz, then we would inevitably end up with `log:V:2` being allowed and working to take the base-2 log of all of the elements of V. And it also seems to mean that `0:(2,6)` would be legal and produce `[0,2,4,6]`, obviously really weird. There are three responses I can think of to this situation: (A) Don't sweat it: encourage use of the `log:(V,2)` form, don't really mention that `log:V:2` would work as well, but don't do anything specifically to suppress that form, and similarly, just tolerate `0:(2,6)` without promoting it; (B) Parse `foo:bar:baz` and `foo:(bar,baz)` into different trees, which use different functions for their semantics, `range` for the first one and say an extended `map` (or a new `elementwise`) for the second. In this scenario, we would _not_ extend the semantics of range, so that `log:V:2` would give a runtime error because log is a function not a number (which is exactly what happens right now). Or (C): `foo:bar` and `foo:(bar)` and `foo:(bar,baz)` all parse to the same basic structure whose semantics are given by a function that covers the binary start:end range case and all elementwise application cases, whereas `foo:bar:baz` parses to a special TernaryRange structure that uses a function for its semantics that only supports the start:step:end case. I think (C) would work perfectly well, but it has the conceptual drawback that it is counterintuitive to coalesce the start:end range and elementwise application semantics while separating the start:step:range, which seems more similar to start:end.

3) `foo:(bar)`

In option B above, this will become very problematic, particularly in the case `foo:(bar+baz)`. Is this a range where the right endpoint is the sum of two numbers, or the elementwise application of a function to the sum of two vectors? This cannot be determined at parse time when the values of the symbols foo, bar, and baz are not known, which seems to imply that `foo:bar` and `foo:(bar)` must both parse to the same sort of node, which in turn seems to rule out option (B) in point (2).

So to sum up all of these points, I see four ways to go at the moment, in no particular order.

(i) Leave things be: mix of non-conflicting operators auto-vectorizing, and needing explicit `map()` for others, possibly combined with improving lambda-expression notation.

(ii) Extend `:` for elementwise operation, tolerate weird cases like `log:V:2` and/or `0:(2,6)`

(iii) Extend `:` for elementwise operation, and avoid those weird cases by having `0:6` and `abs:V` and `log:(V,2)` all parse to similar Node structures while counterintuitively `0:2:6` parses to a different structure. (Note this counter-intuitiveness would only show up to someone who doesn't deal with mathjs syntax trees in one way: there would be one function I'll call here `rangeOrElementwise` so that `rangeOrElementwise(1:4)` is `[1,2,3,4]` while `rangeOrElementwise(log, V, 2)` applies `log` to all of the pairs `(v,2)` for v in V, and a separate function `ternaryRange` that handles `ternaryRange(0,2,6)` to produce `[0, 2, 4, 6]`.)

(iv) Use a different operator symbol than `:(...)` for elementwise function application. The discussion above has suggested `func..(V,W)` or `func@(V,W)`, and there are surely other options; just to brainstorm, doubling the parens `exp((V))`, dollar sign `func$(V,W)`, etc. I think of these I personally like `..` best.

**Comment by dvd101x** [comment]
### Edge case, ranges as arguments of a function

Please consider the case for ranges as inputs in element wise functions with multiple arguments.

``` python
add(1:1:3, 3:-1:1)           # [4, 4, 4]
subtract(1:1:3, 3:-1:1)      # [-2, 0, 2]
dotMultiply(1:1:3, 3:-1:1)   # [3, 4, 3]
dotDivide(1:1:3, 3:-1:1)     # [0.33333333333333, 1, 3]
dotPow(1:1:3, 3:-1:1)        # [1, 4, 3]
```
### Reference from Wolfram Language

I fount in the Wolfram Language docs something that might be useful regarding the discussion for mappings  and element wise functions. I found it interesting as it's Mathematical rather than numerical (like Matlab)

> [Map](https://reference.wolfram.com/language/ref/Map.html)[f,expr] or f/@expr applies f to each element on the first level in expr.
> Evaluate f on each element of a list: `Map[f, {a, b, c, d, e}]`
> Use the short input form: `f /@ {a, b, c, d, e}`

> [Sin](https://reference.wolfram.com/language/ref/Sin.html) can be used with [Interval](https://reference.wolfram.com/language/ref/Interval.html) and [CenteredInterval](https://reference.wolfram.com/language/ref/CenteredInterval.html) objects. [»](https://reference.wolfram.com/language/ref/Sin.html#542663928)
> [Sin](https://reference.wolfram.com/language/ref/Sin.html) automatically threads over lists.
> [Exp](https://reference.wolfram.com/language/ref/Exp.html)[z] gives the exponential of z.
> [Exp](https://reference.wolfram.com/language/ref/Exp.html) automatically threads over lists.
> [Exp](https://reference.wolfram.com/language/ref/Exp.html) can be used with [Interval](https://reference.wolfram.com/language/ref/Interval.html) and [CenteredInterval](https://reference.wolfram.com/language/ref/CenteredInterval.html) objects. [»](https://reference.wolfram.com/language/ref/Exp.html#542663928)
> [MatrixExp](https://reference.wolfram.com/language/ref/MatrixExp.html)[m] gives the matrix exponential of m.

**Comment by gwhitney** [comment]
Right, I think all of these proposals will be compatible with a numeric range as an input, e.g. `log:(1:4, 2)` (but not I think `abs:-3:3`  -- we seem to be leaning toward if you want to do that, you have to write `abs:(-3:3)`, because of the existence of the `start:step:end` notation.

And thanks for the Wolfram reference; it means that if we choose to go with an alternate operator symbol, we could choose `log/@(V,2)` to be similar to Wolfram.

**Comment by josdejong** [comment]
Good analysis Glen! I hadn't realized these side effects of weird cases like `log:V:2` and/or `0:(2,6)` that we get when we use `:` for multiple purposes.

I'm starting to be convinced that we should not reuse `:` for the new map operator: it is technically weird to "overload" the clear `range` function to also be a `map` function, it gives a syntax that allows doing weird things like `0:(2,6)`, and I notice by myself that it is hard to understand, name, and explain this concept when mixing these two concepts together. Then it will most likely be confusing for users too.

Let's see if we can settle on an alternative symbol for the map operator. Thanks David for looking up the syntax that Wolfram uses!

Syntax | Pros | Cons
------ | ---- | ----
~`log:(V,2)` `exp:V`~ | Uses the existing operator `:` for new map like operation | Mixes two concepts together, which is confusing and gives serious complications
`log..(V,2)` `exp..V` | Similar to the element-wise operators `.*`, `./` and `.^` | It may look odd when you see it for the first time?
`log/@(V,2)` `exp/@V` | Same syntax as Mathematica, people may be familiar with it | The `/` feels redundant and doesn't come back in other places of the syntax of mathjs
`log@(V,2)` `exp@V` | Since `@` is used in other languages too related to lambda's and mapping, people may have a natural feel for what it means |  

So far I like both `log..(V,2)` and `log@(V,2)` the best. More thoughts on this?

**Comment by gwhitney** [comment]
Purely in the spirit of brainstorming: the sort of implementation that's being proposed here makes this not really a new operator (infinite notation for a function applied to its left and right operands) but a variant function application notation, that vectorizes rather than applying to entire collections as single entities.

So conceptually this suggests to me that the difference should be some change to the ordinary function-application-notation `f(V,W)`. Since the parentheses are really the only symbols in this notation (no comma in unary case), maybe the notation should be a modification of the parentheses. So the possibilities that come to mind based on this:

1) since mathjs currently doesn't actually allow juxtaposition with a matrix like `3[2,7]` to mean multiplication by  scalar (i.e., producing `[6,21]`), the notation `f[V,W]` actually appears to be available for vectorized application. And the brackets naturally suggest something vector-y going on. The only possible downside I see is if the association with vector-forming is so strong that readers will presume `[V,W]` must mean an array with two entries, each an array itself, and then wonder what it could mean to have a function symbol before it. Other than that potential worry - not sure how real a worry it is - I like this notation; it's certainly briefest and makes it clear this is not an operator but a different sort of function application. 

2) Double the parentheses: `f((V,W))`. The doubling sort of suggests vectorization...

3) Put the modifier character after the open paren: Any of `f(:V,W)`, `f(@V,W)`, `f(*V,W)`, `f(..V,W)`. Can't be `f(.V,W)` because `f(.2,W)` would then be ambiguous between the number .2 and vectorized application to 2 and W. But none of these other four have any prefix meaning, so the notation would work.

4) Like (3) but repeat the decoration symbol at the close paren for symmetry, e.g., `f(:V,W:)`.

5) Use one of these symbols to mark each argument to be vectorized over. This would allow e.g. `log(*V, 2)` for the base 2 log of V and `log(*[4,9,16],*[2,3,4])` to produce `[2,2,2]` but also to disambiguation `pow(M, *[2,3,4]) to mean the array whose entries are the square, cube, and fourth power of (square) matrix M, which would be hard to specify with notations that say only "this is a vectorized function application" without indicating exactly which arguments to vectorize over.

Upon initial reflection, I think I like option (5) best, and I think any of the three symbols `*`, `@`, or `:` would be fine - and there might be other options. I don't think it should be `..` because that's too confusing with splats, whether or not mathjs actually has them.

**Comment by gwhitney** [comment]
And in fact I will say I like (5) with `*` best -- it's like an asterisk indicating "Hey these arguments are being treated specially" and it echoes * in regular expressions indicating sequencing, and * in languages like C that come up in array processing.

**Comment by josdejong** [comment]
That is an interesting idea! So instead of putting a mark on the function, we could put it on the actual argument that we want to iterate over. Conceptually it makes sense. And it gives some interesting new flexibility dealing with element-wise operations. I think though that this concept is a bit more abstract to understand than the element-wise function call like `log@(V, 2)`. And it would not allow this compact `exp@V` notation.

I agree with you that `*` makes sense to use in this case, like `log( *V, 2)` where the `*` is part of the argument and not the function call. In JavaScript the `*` is used to mark a function as a [generator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function*) that you can iterate over, so that is a similar concept. In C and C++ I know it mostly from pointers and de-referencing. These are personal associations, but I get this "watch out complex stuff ahead" warning sign blinking at the inside of my eyes when seeing this `*` in code. Ideally, our element-wise function call does not have associations with being complex, there is no reason for that.

I'll give it a bit more thought 🤔

**Comment by gwhitney** [comment]
> And it would not allow this compact `exp@V` notation.

Weellll, since there is currently no definition of multiplication of a function and a vector, mathjs conceptually _could_ define `square*V` to be the vector of the squares of elements of V, etc. Then you would have this ultracompact notation alongside `log(*V, 2)` and `pow(M,*2:4)`.  In this scenario you could write either `sin(*angles)`  or `sin*angles` -- they would parse differently but produce the same result. I am not necessarily advocating this choice, just exploring the logical possibilities. Nor do I think defining function times collection in this way is a terrible/ridiculous idea, either. Worth contemplating at least.

**Comment by josdejong** [comment]
Ha, that is true. It feels quite confusing to me though to extend `multiply` to do element-wise evaluation when passing it a function as LHS, like it is to extend the function `range` for that purpose. But it is a possibility.

**Comment by gwhitney** [comment]
Well, to my eye, "classic" range and elementwise application were just two distinct operations we were considering gluing together for notational convenience. Here at least, if someone said "Hmmm, we have no definition for the product of a function and a vector, what should that do?" there's not much else that comes to mind other than applying the function to each element of the vector; and we already have some precedent for "special" meanings of * for vectors because of the decision to use `V*W` for the dot product of two vectors. So it feels _less_ of a stretch to me to extend `multiply(func, V)` in this way; it does actually feel legitimately "multiplication-y" to me.

**Comment by gwhitney** [comment]
In any case, `sin(*angles)` is only two characters longer and emphasizes that function application is going on, so it may on the other hand not be necessary to support an ultrashort notation like `sin@angles`. I think you can tell that I like the aspect of the *-argument notation that it makes explicit which argument(s) is/are being vectorized over, and allows you to explicitly _not_ vectorize over a collection argument if you want to simply use that argument repeatedly as a whole, instead of diving in to its elements, but you still want to vectorize over some other argument.

**Comment by josdejong** [comment]
> So it feels _less_ of a stretch to me to extend `multiply(func, V)` in this way; it does actually feel legitimately "multiplication-y" to me.

yeah it's not conflicting in any way. I prefer to not go that direction though. 

Latest thoughts from my side:  I think I prefer `@` over `..` when choosing between those two. Having the ultra compact `exp@V` notation is not _that_ important I think, `exp@(V)` is fine too. 

I do like your proposal of `*` at the argument too, I would like to think that though a bit more to see if there are complications with it or not. Thoughts:

1. What are concrete use cases of these advanced cases where we do want some but not all matrix arguments to be evaluated element-wise, like `pow(M,*2:4)`? I guess it would output an array with matrices. I'm not sure if this is something you would want in practice. If we can only come up with cases where we want all arguments element wise, like `log(*[4,9,16],*[2,3,4])`, then it feels verbose to have to specify the `*` with every argument.
3. Can/should we support the notation for operators too? Like `M ^ *2:4`? I think that is technically possible. So then `A .* B` could be written as `*A * *B`. That looks a bit confusing though, I'm not sure if we should support this.
4. How to implement this? I think we will implement a new function `elementwise`, which as input has a function, a list with arguments, and a list which of the arguments needs to be executed element-wise. The expression parser will create a `FunctionNode` with this `elementwise` function when it encounters the `*` notation.

**Comment by dvd101x** [comment]
I was doing some tests related to numerical integration of ordinary differential equations that could be used in problems like [Rocket Trajectory Optimization](examples/browser/rocket_trajectory_optimization.html) and it seems promising. I think it has the potential to be more versatile than what can be found in numpy or Matlab as here we can also include units and that would make representing physical systems clearer.

The reason I bring this to your attention is that even though I respect the decision to remove the element wise operations to avoid confusion and be more mathematically sound, I think it might affect  numerical computing a bit.

These new ways of mapping functions in the parser are very promising and I was wondering if it's possible to cover both cases? I mean for most functions to accept array likes and also have this versatile ways to map in the parser.

Since the element wise operations were removed I've encountered a few minor issues:

- [Rocket Trajectory Optimization](examples/browser/rocket_trajectory_optimization.html) was expected to work in both ways using `exp` with scalars and also with vectors element wise afterwards
- `cbrt` can't be mapped with the current `map` as the error suggest and it needs a special treatment
- `atan2` is the only trigonometric function that still accepts array likes

**Comment by gwhitney** [comment]
Not sure I am following the question -- `add(...)` requires at least two arguments, as per the following REPL transcript:
```
> M = [[-3, 4], [2, -10]]
   [[-3, 4], [2, -10]]

> add(M)
   TypeError: Too few arguments in function add (expected: any, index: 1)
```

**Comment by josdejong** [comment]
Sorry, bad example. I'll update the question

**Comment by gwhitney** [comment]
On the amended question, my opinion is that since there is no matrix-analytic meaning of the "ceiling of a matrix" other than the element-wise ceiling, I don't see any harm in continuing to allow `ceil(matrix)`. Of course `map(matrix, ceil)` would also work, as `map(matrix, F)` for any unary function `F`. And I guess the `ceil@(matrix)` notation is the final new (accepted?) proposal for trying to make elementwise operations easier in light of some having been removed for the sake of avoiding ambiguity between say the square of a matrix and its elementwise square. But I don't think that's been implemented yet? Presuming it is implemented as a general parsing/operator facility (rather than by simply adding a bunch of new operators named `exp@`, `square@`, etx.) then I don't see any reason why `ceil@(matrix)` should not be allowed to work just like any other `F@(matrix)` would.

**Comment by dvd101x** [comment]
In particular I mean for functions like `sin`, `exp`, etc. that I will call singleton functions.

I encountered a case while implementing a numerical ODE integrator and the function to integrate includes one of those singleton functions. It's common for these methods to use multiple initial values.

```py
y_prime(t, y) = sin(y) + cos(y+1)
initial_state = [1, 1.1]
h = 1
t = 0
next_state = initial_state + h * y_prime(t, initial_state)
```
This throws an error due to the use of a vector inside `sin` and `cos`, but if the function doesn't use the singleton functions it works fine `y_prime(t, y) = t + 2*y + 1`

A workaround is that the user of the numerical ODE integrator defines y_prime function in a map if the function includes any singleton function.
```py
y_prime(t, y) = map(y, sin) + map(y+1, cos)
```
Or maybe it's more legible for it to be 
```py
y_prime(t, y) = map(y, _(y1)=sin(y1)+cos(y1+1))
```
Or manually mapping
```py
y_prime(t, y) = [sin(y[1])+cos(y[1]+1), sin(y[2])+cos(y[2]+1)]
```

With the new map nomenclature it might be more legible, but I don't know outside the parser how it would work.
```py
y_prime(t, y) = sin@(y) + cos@(y+1)
```
This is an example of a possible inconvenience that I encountered during testing in comparison to other numerical ODE integrators.

**Comment by gwhitney** [comment]
Right, there's clearly a need for a function that will be say exp of a number and elementwise exp of a vector. The point of this thread is to settle on what the notation for that should be. I think Jos has settled on `exp@(X)` where `X` can be either a scalar or a vector/matrix. Is there any reason to think that using this notation in the definitions of your y_prime examples wouldn't work (once someone implements it)?

**Comment by josdejong** [comment]
> I think Jos has settled on `exp@(X)` where `X` can be either a scalar or a vector/matrix.

So far I like `exp@(X)` quite well and know that is technically feasible, but I would like to explore Glen's idea of `log( *V, 2)` further (see https://github.com/josdejong/mathjs/issues/2440#issuecomment-1419139918, https://github.com/josdejong/mathjs/issues/2440#issuecomment-1433368026), see what would be involved to implement that, and think through the use cases.

**Comment by dvd101x** [comment]
I was reading a similar discussion in the Julia community.

- https://discourse.julialang.org/t/what-is-the-future-of-sin-vector-when-there-is-sin-vector/677/19
- https://stackoverflow.com/questions/59232420/how-does-one-perform-the-exp-operation-element-wise-in-juila

After reading that I’m detracting from my request to allow both methods. It seems to work well for them and cover more cases.

If anything I would only question if in the meantime of this element wise operator is implemented you might consider returning back the old matlab style behavior. Now that `solveODE` is implemented I could show you some specific cases for y_prime that might seem inconvenient if that’s ok.

**Comment by gwhitney** [comment]
Thanks for these references. I think they do reinforce where this discussion seemed to end up, which is for mathjs not to be concerned with making raw functions work on both scalars and matrices, but instead to provide convenient notation for vectorizing any function.  I just am not sure if that notation ever quite settled...

**Comment by dvd101x** [comment]
Thanks for the review Glen,

Today I was reading in detail the blogpost in which they announced the change they made which is very inline with this discussion.

https://julialang.org/blog/2017/01/moredots/

In particular I found interesting they use some of the characters mentioned here.

`X .= f.(2 .* X.^2 .+ 6 .* X.^3 .- sqrt.(X))`

`@. X = f(2X^2 + 6X^3 - sqrt(X))`

And also their description of broadcast and how it would apply for functions with multiple arguments.

Something interesting is that for Julia the `.` is the broadcast operator and is mandatory for something like this.

```jl 
f(a,b) = a+b
f.([1,2],[3,4]')
```

Which in Matlab, Numpy and `mathjs` is not mandatory.

**Comment by gwhitney** [comment]
Since the discussion here has shifted entirely to a broadcast operator for mathjs (and the original name of the issue has long since happened), renaming the issue.

**Comment by dvd101x** [comment]
Also now it's possible to map multiple arguments and broadcasting them, so if the operator is implemented with multiple arguments in mind, now the map function can support that.

**Comment by gwhitney** [comment]
Happened to come back and glance at this today. Since we already use `.*`, `.^`, and `.+`, is there any argument against dropping those operators and just adopting the Julia broadcasting conventions wholesale into Julia? The only question I can think of is whether parses  of `.` would become too ambiguous  but I don't see that as an issue, in that the "broadcast dots" are limited to just-before-an operator, where dots currently cannot occur,  and to between-a-function-and-its-arguments, where again a dot cannot currently occur. While `f?..(M)` for optionally chaining a broadcast function call will look a little weird, I think it is unambiguous (and probably won't come up that much). So after this long hiatus, that would be my strong vote here. It avoids needing to use a new character (unless we also adopted Julia's `@.` notation for "broadcast everything in the following expression, but I'd suggest if we get to the point we want that we actually use a different notation to keep `@` available for other possible uses). And it completely obviates the whole `log(*V, 2)` proposal, in that if `log.(V, 2)` broadcasts `log` over V and 2, it will do just what we wanted `log(*V, 2)` to do.  I think imitating the best convention for broadcasting out there should become part of the mathjs roadmap. (I think we should have a roadmap ;-)

**Comment by dvd101x** [comment]
Hi Glen, those operators (`.*`, `.^` and `.+`) are also available in Julia, I don't know if you mean that they would become hard to parse with all the other uses of `.`. For me keeping those operators and adding the `sin.(M)` and `log.(V, 2)` is ideal. The `@` syntax is nice, but I don't fully understand the other uses of `@` available in Julia, maybe it could be a decision for later.

Syntax wise I think Julia is the closest to mathjs vision, not everything but some references can be still be taken from Matlab, Python and Mathematica.

**Comment by gwhitney** [comment]
Sorry I guess I am a little confused  -- are .*, ./, and .^ their own operators in Julia, or do they work the way they do just as consequences of the general . broadcast notation?
